internal/sysfs/select_windows.go

```diff
@@ -37,69 +33,181 @@ func syscall_select(n int, r, w, e *platform.FdSet, timeout *time.Duration) (int
 		time.Sleep(*timeout)
 		return 0, nil
 	}
-	if r.IsSet(wasiFdStdin) {
-		fileType, err := syscall.GetFileType(syscall.Stdin)
-		if err != nil {
-			return 0, err
-		}
-		if fileType&syscall.FILE_TYPE_CHAR != 0 {
-			res, err := pollNamedPipe(context.TODO(), syscall.Stdin, timeout)
-			if err != nil {
-				return -1, err
-			}
-			if !res {
-				r.Zero()
-				return 0, nil
-			}
-		}
-		r.Zero()
-		r.Set(wasiFdStdin)
-		return 1, nil
+
+	n, errno := selectAllHandles(context.TODO(), r, w, e, timeout)
+	if errno == 0 {
+		return n, nil
 	}
-	return -1, sys.ENOSYS
+	return n, errno
 }
 
-// pollNamedPipe polls the given named pipe handle for the given duration.
+// selectAllHandles emulates a general-purpose POSIX select on Windows.
 //
 // The implementation actually polls every 100 milliseconds until it reaches the given duration.
 // The duration may be nil, in which case it will wait undefinely. The given ctx is
-// used to allow for cancellation. Currently used only in tests.
-func pollNamedPipe(ctx context.Context, pipeHandle syscall.Handle, duration *time.Duration) (bool, error) {
+// used to allow for cancellation, and it is currently used only in tests.
+//
+// As indicated in the man page for select [1], r, w, e are modified upon completion:
+//
+// "Upon successful completion, the pselect() or select() function shall modify the objects pointed to by the readfds,
+// writefds, and errorfds arguments to indicate which file descriptors are ready for reading, ready for writing,
+// or have an error condition pending, respectively, and shall return the total number of ready descriptors in all the output sets"
```

Comment:
As this is emulation with goroutines, it is worthwhile to explain what is in windows, but not possible to use. Also, any downside such as this implies GOMAXPROCS>1 if that's true. This explanation could be in the RATIONALE.md like "Why do we emulate POSIX select on windows?"

Reply:
💯 on adding a section to RATIONALE. The tl;dr: is that I don't think there is one single API that does everything `select` does on POSIX, and, because winsock select is blocking under certain conditions we need to spin a goroutine to avoid blocking everything

another important note: TBH even UNIX select is not perfect for our use cases, because it won't honor the clock config of the module; so if you have a "poll 1 sec fd 4", and you implemented it with `select`, you would end up blocking for literally 1 sec. So if you wanted to do it properly, I think you may have to use some concurrency primitives there as well.

OTOH when this type of I/O is involved I don't even know if it composes with simulated clocks at all.

Reply:
agree on the tension of real clocks and real files. I found myself in that spaghetti before, thanks for clarifying!

Reply:
thank you!